File: crud-item-list/src/test/java/com/gurunars/crud_item_list/MoverDownTest.java
package com.gurunars.crud_item_list;

import org.junit.Test;
import org.mockito.internal.util.collections.Sets;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import static junit.framework.Assert.assertEquals;

public class MoverDownTest {

    private final ActionMoveDown<String> moverDown = new ActionMoveDown<>();
    private final List<String> all = Arrays.asList(
        "one",
        "two",
        "three",
        "four",
        "five",
        "six",
        "seven",
        "eight",
        "nine",
        "ten"
    );
    private final Set<String> selected = Sets.newSet(
        "four",
        "five",
        "six",
        "seven"
    );

    private final List<String> expectedOutcome = Arrays.asList(
        "one",
        "two",
        "three",
        "eight", // item moved up
        "four",
        "five",
        "six",
        "seven",
        "nine",
        "ten"
    );

    private List<String> apply(Set<String> selected) {
        List<String> result = new ArrayList<>(all);
        moverDown.perform(result, selected);
        return result;
    }

    @Test
    public void moveDown_isCorrect() throws Exception {
        assertEquals(expectedOutcome, apply(selected));
    }

}

File: crud-item-list/src/test/kotlin/com/gurunars/crud_item_list/MoverDownTest.kt
package com.gurunars.crud_item_list

import org.junit.Assert.assertEquals
import org.junit.Test

class MoverDownTest {

    private val moverDown = ActionMoveDown<StringItem>()
    private val all = listOf(
        "one",
        "two",
        "three",
        "four",
        "five",
        "six",
        "seven",
        "eight",
        "nine",
        "ten"
    )
    private val selected = setOf(
        "four",
        "five",
        "six",
        "seven"
    )

    private val expectedOutcome = listOf(
        "one",
        "two",
        "three",
        "eight", // item moved up
        "four",
        "five",
        "six",
        "seven",
        "nine",
        "ten"
    )

    @Test
    @Throws(Exception::class)
    fun moveDown_isCorrect() {
        moverDown.perform(all, selected, { all -> assertEquals(expectedOutcome, all) })
    }

}
